File: haskell/src/Surd/Radical/Expr.hs
-- |
-- Module      : Surd.Radical.Expr
-- Description : Smart constructors and structural queries for radical expressions
-- Stability   : experimental
--
-- Convenience operations for building and inspecting 'RadExpr' trees.
-- These functions do /not/ normalize -- they just construct the AST nodes.
-- Use "Surd.Radical.Normalize" for simplification.
module Surd.Radical.Expr
  ( -- * Construction helpers
    lit,
    neg,
    add,
    sub,
    mul,
    div',
    inv,
    root,
    sqrt',
    pow,
    fromInteger',
    fromRational',

    -- * Structural queries
    depth,
    size,
    freeOf,
    mapCoeffs,

    -- * Radical collection and dependency ordering
    collectRadicals,
    topoSortRadicals,
    allRootsResolved,
  )
where

import Data.List (nub)
import Surd.Types
import Prelude hiding (div, sqrt)

-- | Build a coefficient literal node.
lit :: k -> RadExpr k
lit = Lit

-- | Build a negation node.
neg :: RadExpr k -> RadExpr k
neg = Neg

-- | Build an addition node.
add :: RadExpr k -> RadExpr k -> RadExpr k
add = Add

-- | Build a subtraction node: @sub a b = Add a (Neg b)@.
sub :: RadExpr k -> RadExpr k -> RadExpr k
sub a b = Add a (Neg b)

-- | Build a multiplication node.
mul :: RadExpr k -> RadExpr k -> RadExpr k
mul = Mul

-- | Build a division node: @div' a b = Mul a (Inv b)@.
--
-- Named @div'@ to avoid clashing with 'Prelude.div'.
div' :: RadExpr k -> RadExpr k -> RadExpr k
div' a b = Mul a (Inv b)

-- | Build a multiplicative inverse node.
inv :: RadExpr k -> RadExpr k
inv = Inv

-- | Build an @n@th root node: @root n x@ represents @x^(1/n)@.
root :: Int -> RadExpr k -> RadExpr k
root = Root

-- | Build a square root node: @sqrt' x = Root 2 x@.
--
-- Named @sqrt'@ to avoid clashing with 'Prelude.sqrt'.
sqrt' :: RadExpr k -> RadExpr k
sqrt' = Root 2

-- | Build an integer power node: @pow e n@ represents @e^n@.
pow :: RadExpr k -> Int -> RadExpr k
pow = Pow

-- | Lift an 'Integer' into a @RadExpr Rational@.
fromInteger' :: Integer -> RadExpr Rational
fromInteger' = Lit . fromInteger

-- | Lift a 'Rational' into a @RadExpr Rational@.
fromRational' :: Rational -> RadExpr Rational
fromRational' = Lit

-- | Nesting depth of the expression tree.
--
-- Leaves ('Lit') have depth 0. Each unary or binary node adds 1.
depth :: RadExpr k -> Int
depth (Lit _) = 0
depth (Neg a) = depth a
depth (Add a b) = 1 + max (depth a) (depth b)
depth (Mul a b) = 1 + max (depth a) (depth b)
depth (Inv a) = 1 + depth a
depth (Root _ a) = 1 + depth a
depth (Pow a _) = 1 + depth a

-- | Number of nodes in the expression tree (including leaves).
size :: RadExpr k -> Int
size (Lit _) = 1
size (Neg a) = 1 + size a
size (Add a b) = 1 + size a + size b
size (Mul a b) = 1 + size a + size b
size (Inv a) = 1 + size a
size (Root _ a) = 1 + size a
size (Pow a _) = 1 + size a

-- | Test whether all coefficients in the expression satisfy a predicate.
--
-- This traverses the entire tree, returning 'True' only if the predicate
-- holds for every 'Lit' coefficient. Despite the name, it does not
-- specifically check for 'Root' nodes.
freeOf :: (k -> Bool) -> RadExpr k -> Bool
freeOf p (Lit k) = p k
freeOf p (Neg a) = freeOf p a
freeOf p (Add a b) = freeOf p a && freeOf p b
freeOf p (Mul a b) = freeOf p a && freeOf p b
freeOf p (Inv a) = freeOf p a
freeOf p (Root _ a) = freeOf p a
freeOf p (Pow a _) = freeOf p a

-- | Map a function over the coefficients. Same as 'fmap' but with
-- a more descriptive name for the radical expression domain.
mapCoeffs :: (a -> b) -> RadExpr a -> RadExpr b
mapCoeffs = fmap

-- | Collect distinct @(rootIndex, radicand)@ pairs from an expression.
--
-- Traverses the tree, collecting every 'Root' occurrence as a pair
-- @(n, radicand)@. Duplicates are removed (preserving first-occurrence
-- order via 'nub').
collectRadicals :: (Eq k) => RadExpr k -> [(Int, RadExpr k)]
collectRadicals = nub . go
  where
    go (Lit _) = []
    go (Neg a) = go a
    go (Add a b) = go a ++ go b
    go (Mul a b) = go a ++ go b
    go (Inv a) = go a
    go (Pow a _) = go a
    go (Root n a) = go a ++ [(n, a)]

-- | Topologically sort radicals so that radicals with rational radicands
-- come first, followed by radicals whose radicands depend only on
-- earlier radicals.
--
-- If cyclic dependencies exist (which should not happen for well-formed
-- radical expressions), unresolvable radicals are appended at the end.
topoSortRadicals :: (Eq k) => [(Int, RadExpr k)] -> [(Int, RadExpr k)]
topoSortRadicals = go []
  where
    go sorted [] = sorted
    go sorted remaining =
      let ready = [r | r <- remaining, allRootsResolved sorted (snd r)]
          remaining' = filter (`notElem` ready) remaining
       in if null ready
            then sorted ++ remaining -- can't resolve more; append as-is
            else go (sorted ++ ready) remaining'

-- | Check whether all 'Root' subexpressions in a radicand are present
-- in the resolved set. Used by 'topoSortRadicals' to determine which
-- radicals are ready to be processed next.
allRootsResolved :: (Eq k) => [(Int, RadExpr k)] -> RadExpr k -> Bool
allRootsResolved _ (Lit _) = True
allRootsResolved resolved (Neg a) = allRootsResolved resolved a
allRootsResolved resolved (Add a b) = allRootsResolved resolved a && allRootsResolved resolved b
allRootsResolved resolved (Mul a b) = allRootsResolved resolved a && allRootsResolved resolved b
allRootsResolved resolved (Inv a) = allRootsResolved resolved a
allRootsResolved resolved (Pow a _) = allRootsResolved resolved a
allRootsResolved resolved (Root n a) = (n, a) `elem` resolved

File: scala3/src/main/scala/surd/Expr.scala
package surd

import RadExpr.*

/** Convenience constructors and structural queries on radical expressions.
  *
  * These do NOT normalise -- they just build or inspect the AST.
  */
object Expr:

  // ---------------------------------------------------------------
  // Construction helpers
  // ---------------------------------------------------------------

  /** Literal coefficient. */
  def lit[K](v: K): RadExpr[K] = Lit(v)

  def neg[K](e: RadExpr[K]): RadExpr[K] = Neg(e)

  def add[K](a: RadExpr[K], b: RadExpr[K]): RadExpr[K] = Add(a, b)

  def sub[K](a: RadExpr[K], b: RadExpr[K]): RadExpr[K] = Add(a, Neg(b))

  def mul[K](a: RadExpr[K], b: RadExpr[K]): RadExpr[K] = Mul(a, b)

  def div[K](a: RadExpr[K], b: RadExpr[K]): RadExpr[K] = Mul(a, Inv(b))

  def inv[K](e: RadExpr[K]): RadExpr[K] = Inv(e)

  def root[K](n: Int, e: RadExpr[K]): RadExpr[K] = Root(n, e)

  def sqrt[K](e: RadExpr[K]): RadExpr[K] = Root(2, e)

  def pow[K](e: RadExpr[K], n: Int): RadExpr[K] = Pow(e, n)

  def fromInt(n: Int): RadExpr[Rational] = Lit(Rational(n))

  def fromBigInt(n: BigInt): RadExpr[Rational] = Lit(Rational(n))

  def fromRational(r: Rational): RadExpr[Rational] = Lit(r)

  // ---------------------------------------------------------------
  // Structural queries
  // ---------------------------------------------------------------

  /** Nesting depth of the expression tree. */
  def depth[K](e: RadExpr[K]): Int = e match
    case Lit(_)      => 0
    case Neg(a)      => depth(a)
    case Add(a, b)   => 1 + math.max(depth(a), depth(b))
    case Mul(a, b)   => 1 + math.max(depth(a), depth(b))
    case Inv(a)      => 1 + depth(a)
    case Root(_, a)  => 1 + depth(a)
    case Pow(a, _)   => 1 + depth(a)

  /** Number of nodes in the expression tree. */
  def size[K](e: RadExpr[K]): Int = e match
    case Lit(_)      => 1
    case Neg(a)      => 1 + size(a)
    case Add(a, b)   => 1 + size(a) + size(b)
    case Mul(a, b)   => 1 + size(a) + size(b)
    case Inv(a)      => 1 + size(a)
    case Root(_, a)  => 1 + size(a)
    case Pow(a, _)   => 1 + size(a)

  /** Check whether every coefficient in the expression satisfies predicate p. */
  def freeOf[K](p: K => Boolean)(e: RadExpr[K]): Boolean = e match
    case Lit(k)      => p(k)
    case Neg(a)      => freeOf(p)(a)
    case Add(a, b)   => freeOf(p)(a) && freeOf(p)(b)
    case Mul(a, b)   => freeOf(p)(a) && freeOf(p)(b)
    case Inv(a)      => freeOf(p)(a)
    case Root(_, a)  => freeOf(p)(a)
    case Pow(a, _)   => freeOf(p)(a)

  /** Map a function over the coefficients. */
  def mapCoeffs[A, B](f: A => B)(e: RadExpr[A]): RadExpr[B] = RadExpr.map(e)(f)

  // ---------------------------------------------------------------
  // Radical collection and dependency ordering
  // ---------------------------------------------------------------

  /** Collect distinct (rootIndex, radicand) pairs from an expression.
    * Uses structural equality for deduplication.
    */
  def collectRadicals[K](e: RadExpr[K]): List[(Int, RadExpr[K])] =
    go(e).distinct

  private def go[K](e: RadExpr[K]): List[(Int, RadExpr[K])] = e match
    case Lit(_)      => Nil
    case Neg(a)      => go(a)
    case Add(a, b)   => go(a) ++ go(b)
    case Mul(a, b)   => go(a) ++ go(b)
    case Inv(a)      => go(a)
    case Pow(a, _)   => go(a)
    case Root(n, a)  => go(a) :+ (n, a)

  /** Topologically sort radicals so that radicals with rational radicands
    * come first, followed by radicals whose radicands depend only on
    * earlier radicals. Unresolvable radicals are appended at the end.
    */
  def topoSortRadicals[K](rads: List[(Int, RadExpr[K])]): List[(Int, RadExpr[K])] =
    def loop(sorted: List[(Int, RadExpr[K])], remaining: List[(Int, RadExpr[K])]): List[(Int, RadExpr[K])] =
      if remaining.isEmpty then sorted
      else
        val ready = remaining.filter(r => allRootsResolved(sorted, r._2))
        val rest = remaining.filterNot(ready.contains)
        if ready.isEmpty then sorted ++ remaining  // can't resolve more
        else loop(sorted ++ ready, rest)
    loop(Nil, rads)

  /** Check whether all Root subexpressions in a radicand are present
    * in the resolved set.
    */
  def allRootsResolved[K](resolved: List[(Int, RadExpr[K])], e: RadExpr[K]): Boolean = e match
    case Lit(_)      => true
    case Neg(a)      => allRootsResolved(resolved, a)
    case Add(a, b)   => allRootsResolved(resolved, a) && allRootsResolved(resolved, b)
    case Mul(a, b)   => allRootsResolved(resolved, a) && allRootsResolved(resolved, b)
    case Inv(a)      => allRootsResolved(resolved, a)
    case Pow(a, _)   => allRootsResolved(resolved, a)
    case Root(n, a)  => resolved.contains((n, a))
